kem/frodo/frodo640shake/frodo.go

```diff
@@ -0,0 +1,623 @@
+	}
+	return nil
+}
+
+// Packs sk to buf.
+//
+// Panics if buf is not of size PrivateKeySize.
+func (sk *PrivateKey) Pack(buf []byte) {
+	if len(buf) != PrivateKeySize {
+		panic("buf must be of length PrivateKeySize")
+	}
+
+	copy(buf[:SharedKeySize], sk.hashInputIfDecapsFail[:])
+	buf = buf[SharedKeySize:]
+
+	sk.pk.Pack(buf[:PublicKeySize])
+	buf = buf[PublicKeySize:]
+
+	j := 0
+	for i := range sk.matrixS {
+		buf[j] = byte(sk.matrixS[i])
+		buf[j+1] = byte(sk.matrixS[i] >> 8)
+		j += 2
+	}
+	buf = buf[j:]
+
+	copy(buf[:], sk.hpk[:])
+}
+
+// Unpacks sk from buf.
+//
+// Panics if buf is not of size PrivateKeySize.
+func (sk *PrivateKey) Unpack(buf []byte) {
+	if len(buf) != PrivateKeySize {
+		panic("buf must be of length PrivateKeySize")
+	}
+
+	copy(sk.hashInputIfDecapsFail[:], buf[:SharedKeySize])
+	buf = buf[SharedKeySize:]
+
+	sk.pk = new(PublicKey)
+	sk.pk.Unpack(buf[:PublicKeySize])
+	buf = buf[PublicKeySize:]
+
+	for i := range sk.matrixS {
+		sk.matrixS[i] = uint16(buf[i*2]) | (uint16(buf[(i*2)+1]) << 8)
+	}
+	buf = buf[len(sk.matrixS)*2:]
+
+	copy(sk.hpk[:], buf[:])
+}
+
+// Packs pk to buf.
+//
+// Panics if buf is not of size PublicKeySize.
+func (pk *PublicKey) Pack(buf []byte) {
+	if len(buf) != PublicKeySize {
+		panic("buf must be of length PublicKeySize")
+	}
+
+	copy(buf[:seedASize], pk.seedA[:])
+	pack(buf[seedASize:], pk.matrixB[:])
+}
+
+// TODO: Unpacks pk from buf.
+//
+// Panics if buf is not of size PublicKeySize.
+func (pk *PublicKey) Unpack(buf []byte) {
+	if len(buf) != PublicKeySize {
+		panic("buf must be of length PublicKeySize")
+	}
+
+	copy(pk.seedA[:], buf[:seedASize])
+	unpack(pk.matrixB[:], buf[seedASize:])
+}
+
+// Boilerplate down below for the KEM scheme API.
+
+type scheme struct{}
+
+var sch kem.Scheme = &scheme{}
+
+// Scheme returns a KEM interface.
+func Scheme() kem.Scheme { return sch }
+
+func (scheme) Name() string                { return "FrodoKEM-640-SHAKE" }
+func (*scheme) PublicKeySize() int         { return PublicKeySize }
+func (*scheme) PrivateKeySize() int        { return PrivateKeySize }
+func (*scheme) SeedSize() int              { return KeySeedSize }
+func (*scheme) SharedKeySize() int         { return SharedKeySize }
+func (*scheme) CiphertextSize() int        { return CiphertextSize }
+func (*scheme) EncapsulationSeedSize() int { return EncapsulationSeedSize }
+
+func (sk *PrivateKey) Scheme() kem.Scheme { return sch }
+func (pk *PublicKey) Scheme() kem.Scheme  { return sch }
+
+func (sk *PrivateKey) MarshalBinary() ([]byte, error) {
+	var ret [PrivateKeySize]byte
+	sk.Pack(ret[:])
+	return ret[:], nil
+}
+
+func (sk *PrivateKey) Equal(other kem.PrivateKey) bool {
+	oth, ok := other.(*PrivateKey)
+	if !ok {
+		return false
+	}
+	if sk.pk == nil && oth.pk == nil {
+		return true
+	}
+	if sk.pk == nil || oth.pk == nil {
+		return false
+	}
+	return ctCompareU16(sk.matrixS[:], oth.matrixS[:]) == 0 &&
+		bytes.Equal(sk.hashInputIfDecapsFail[:], oth.hashInputIfDecapsFail[:]) &&
```

Comment:
better to use ConstantTimeCompare
 https://pkg.go.dev/crypto/subtle#ConstantTimeCompare